### PR TITLE
feat: contracts and project providers + cleanup

### LINKED ISSUE #626: `substrate-contracts-node` provider
[substrate-contracts-node](https://github.com/paritytech/substrate-contracts-node) provider

### PR DESCRIPTION
**Pull request by harrysolovay**
Resolves #626

This PR introduces two new providers:

1. A `ProjectProvider`, which allows easy use within a substrate project. Simply `capi -- deno run -A my_script.ts` in your Substrate project.

`my_script.ts`

```ts
import { MyCustomPallet } from "http://localhost:4646/frame/project/@latest/mod.ts"
import { assertEquals } from "https://deno.land/std/testing/asserts.ts"

Deno.test("works", async () => {
  assertEquals(expected, await MyCustomPallet.item.run())
})
```

2. A contracts provider, which wraps [`substrate-contracts-node`](https://github.com/paritytech/substrate-contracts-node) (per the request of @wottpal).

```ts
import { ... } from "http://localhost:4646/frame/contracts_dev/@latest/mod.ts"
```

---

This PR also...

- extracts common provider functionality into an abstract `FrameBinProvider` class
- changes the shape in which provider factories are passed to `Env`
- improved implementation of test provider launch timeouts

---

Next up: #649

### CODE REVIEW

**Comment by kratico** [comment]
`substrate-contracts-node` does not emit a `author_extrinsicUpdate` finalized event with `result: { finalized: "0x..." }`

Perhaps, for instant finality, the `inBlock` event should be the terminal transaction status.

The following is a work around for local debugging

```
diff --git a/fluent/ExtrinsicStatusRune.ts b/fluent/ExtrinsicStatusRune.ts
index 52b6c281..02638cb7 100644
--- a/fluent/ExtrinsicStatusRune.ts
+++ b/fluent/ExtrinsicStatusRune.ts
@@ -55,7 +55,7 @@ export class ExtrinsicStatusRune<out U1, out U2, out C extends Chain = Chain>
   }
 
   txEvents() {
-    const block = this.finalized()
+    const block = this.inBlock()
     const txI = Rune
       .tuple([
         block.into(ValueRune).access("block", "extrinsics"),
diff --git a/rpc/known/author.ts b/rpc/known/author.ts
index b41f5297..b107ee84 100644
--- a/rpc/known/author.ts
+++ b/rpc/known/author.ts
@@ -84,7 +84,7 @@ export namespace TransactionStatus {
     return typeof inQuestion === "string"
       ? inQuestion === "invalid" || inQuestion === "dropped"
       : !!(inQuestion.finalized || inQuestion.finalityTimeout || inQuestion.retracted
-        || inQuestion.usurped)
+        || inQuestion.usurped || inQuestion.inBlock)
   }
 }
```

**Comment by harrysolovay** [comment]
@kratico thank you for the extraordinary contracts fix! The contracts example now executes in less than a second!

File: server/local.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ export function handler(env: Env): Handler {
           return await provider.handle(request, pathInfo)
         } catch (e) {
           if (e instanceof Response) return e
-          if (env.dbg) console.error(e)
+          console.error(e)
```
Comment (harrysolovay) [inline server/local.ts:26]: We can re-add the toggling on/off when we tackle a comprehensive approach to logging from the Capi server (perhaps using `Logger` from `deno_std`). I plan to take a pass at this in a PR that also addresses #649.

**Comment by wottpal** [comment]
> `substrate-contracts-node` does not emit a `author_extrinsicUpdate` finalized event with `result: { finalized: "0x..." }`
> 
> Perhaps, for instant finality, the `inBlock` event should be the terminal transaction status.
> 
> …

Just for reference, this would probably also fix this implicitly: https://github.com/paritytech/substrate-contracts-node/issues/160

**Review by tjjfvi — APPROVED** [review]

**Review by tjjfvi — APPROVED** [review]

**Pull request merged**